File: O(nLogn)/merge_sort.py
def merge_sort(arr):
    # ---------------------------------------------------------
    # PARTE 1: O(log n) - A Divisão
    # A recursão divide o array ao meio sucessivamente.
    # Quantas vezes podemos dividir N por 2 até chegar em 1?
    # Resposta: log₂(n) vezes. Isso define a "altura" da árvore de chamadas.
    # ---------------------------------------------------------
    if len(arr) > 1:
        meio = len(arr) // 2
        esquerda = arr[:meio]
        direita = arr[meio:]

        # Chamada recursiva para cada metade
        merge_sort(esquerda)
        merge_sort(direita)

        # -----------------------------------------------------
        # PARTE 2: O(n) - A Conquista (Intercalação)
        # Aqui ocorre o trabalho real de ordenar e juntar.
        # Em CADA nível da árvore de recursão (do topo à base),
        # nós percorremos e ordenamos TODOS os n elementos.
        # -----------------------------------------------------
        i = j = k = 0

        # O laço while percorre os elementos das sub-listas.
        # Somando todas as iterações deste laço em um mesmo nível de profundidade,
        # o total de operações é proporcional a n.
        while i < len(esquerda) and j < len(direita):
            if esquerda[i] < direita[j]:
                arr[k] = esquerda[i]
                i += 1
            else:
                arr[k] = direita[j]
                j += 1
            k += 1

        # Os laços abaixo copiam os elementos restantes.
        # Eles fazem parte do custo linear O(n) deste nível.
        while i < len(esquerda):
            arr[k] = esquerda[i]
            i += 1
            k += 1

        while j < len(direita):
            arr[k] = direita[j]
            j += 1
            k += 1

'''
------------------------------------------------------------------
COMENTÁRIO COM O CÁLCULO DA COMPLEXIDADE (PARA O TCC):

A complexidade final é o produto da "Altura da Árvore" pela "Largura do Trabalho".

1. Altura (Níveis de Recursão):
   O algoritmo divide n por 2 repetidamente.
   Ex: 8 -> 4 -> 2 -> 1.
   Isso gera uma profundidade de log₂ n.
   -> Custo: log n

2. Largura (Trabalho por Nível):
   Em cada nível da árvore, o processo de 'merge' (os whiles)
   precisa iterar sobre todos os elementos para juntá-los.
   Independente se são 2 listas de n/2 ou 4 listas de n/4,
   a soma total de elementos processados naquele nível é n.
   -> Custo: n

3. Cálculo Final:
   Total = (Trabalho por Nível) * (Número de Níveis)
   Total = (n) * (log n)
   Resultado = O(n log n)
------------------------------------------------------------------
'''

# Teste
numeros = [12, 11, 13, 5, 6, 7]
merge_sort(numeros)
print(f"Array ordenado: {numeros}")

File: O(nLogn)/MergeSort.java
public class MergeSort {
    
    public static void sort(int arr[], int l, int r) {
        // ---------------------------------------------------------
        // PARTE 1: O(log n) - A Divisão
        // A recursão cria uma árvore binária de chamadas.
        // A profundidade (altura) dessa árvore é determinada por quantas
        // vezes conseguimos dividir n por 2: exatamente log n vezes.
        // ---------------------------------------------------------
        if (l < r) {
            int m = l + (r - l) / 2; 

            sort(arr, l, m);
            sort(arr, m + 1, r);

            // Chama a função que realiza o trabalho linear
            merge(arr, l, m, r);
        }
    }

    public static void merge(int arr[], int l, int m, int r) {
        int n1 = m - l + 1;
        int n2 = r - m;

        int E[] = new int[n1];
        int D[] = new int[n2];

        for (int i = 0; i < n1; ++i) E[i] = arr[l + i];
        for (int j = 0; j < n2; ++j) D[j] = arr[m + 1 + j];

        int i = 0, j = 0;
        int k = l; 

        // ---------------------------------------------------------
        // PARTE 2: O(n) - A Conquista (Intercalação)
        // Independentemente de estarmos no topo ou na base da recursão,
        // a soma das iterações deste while para um nível inteiro da árvore
        // será proporcional a n (tamanho total do array).
        // ---------------------------------------------------------
        while (i < n1 && j < n2) {
            if (E[i] <= D[j]) {
                arr[k] = E[i];
                i++;
            } else {
                arr[k] = D[j];
                j++;
            }
            k++;
        }

        while (i < n1) arr[k++] = E[i++];
        while (j < n2) arr[k++] = D[j++];
    }

    /*
    ------------------------------------------------------------------
    CÁLCULO DA COMPLEXIDADE:

    O Merge Sort aplica a estratégia "Dividir para Conquistar".

    1. Custo da Divisão (Recursão): 
       O problema é dividido em subproblemas de tamanho n/2.
       Isso gera uma altura de árvore de log n.
       
    2. Custo da Conquista (Merge):
       Em cada nível da recursão, o algoritmo precisa percorrer e copiar
       todos os n elementos para seus lugares ordenados.
       Custo por nível: O(n).

    3. Resultado Matemático:
       Complexidade = O(n) operações repetidas O(log n) vezes.
       Total = O(n log n).
    ------------------------------------------------------------------
    */

    public static void main(String args[]) {
        int arr[] = {12, 11, 13, 5, 6, 7};
        sort(arr, 0, arr.length - 1);
        System.out.println("Array ordenado");
        for (int n : arr) System.out.print(n + " ");
    }
}